src/NUnitFramework/framework/Constraints/CollectionOrderedConstraint.cs

```diff
@@ -103,7 +124,12 @@ public CollectionOrderedConstraint Using<T>(Comparison<T> comparer)
         /// </summary>
         public CollectionOrderedConstraint By(string propertyName)
         {
-            this.propertyName = propertyName;
+            if (!this.allowBy) 
+                throw new InvalidOperationException("The use of 'By' is not allowed.");
+
+            this.properties.Add(new OrderedPropertyInfo(propertyName, this.descending));
+            this.allowBy = false;
```

Comment:
I think this is incorrect. At least I don't see anything wrong with Ordered.By.By. However Ordered.By.Descending.By should be wrong. IMO, you should entirely remove the allowBy thing and get this working first without that error check. Once we have the right overall shape, then we can deal with the edge case.